### PR TITLE
Optimise for fixed path prefixes in glob()

### PR DESCRIPTION
**Pull request by peterebden**
Basically lets us walk into fixed prefix parts efficiently rather than always walking over every possible path and trying to match them.

Reduces my test case from 5s -> 0.5s

Mildly worried about the increase in complexity here. All tests pass but I am a little worried that it'd be possible to get some edge case wrong.

### CODE REVIEW

**Review by Tatskaari — APPROVED** [review]
This this makes sense. If we can basically build core3 with this then I'm happy.

**Comment by peterebden** [comment]
Cool - I might experiment with it locally a bit more. Will have a look internally once we get up to a recent prerelease.

File: src/fs/glob.go [hunk 1]
```diff
@@ -174,10 +240,26 @@ func (globber *Globber) walkDir(rootPath string) (walkedDir, error) {
 	return dir, nil
 }
 
+// globPrefix extracts the fixed prefix part from a glob expression.
+// i.e. "src/fs/**/*.go" -> ["src", "fs"], ["**", "/*.go"]
```
Comment (tiagovtristao) [inline src/fs/glob.go:244]: nit: The last bit tripped me. I think you meant just `// i.e. "src/fs/**/*.go" -> ["src", "fs"]`?

File: src/fs/glob.go [hunk 2]
```diff
@@ -76,7 +77,67 @@ type Globber struct {
...
+	child, present := ft.children[prefix[0]]
+	if !present {
+		return nil
+	}
+	files := child.AllPrefixed(prefix[1:])
+	ret := make([]string, len(files))
+	for i, name := range files {
+		ret[i] = path.Join(prefix[0], name)
+	}
+	sort.Strings(ret)
```
[the remark this answers is not in the document]
Comment (peterebden) [reply src/fs/glob.go:116]: Ah good spot!

File: src/fs/glob.go (hunk 1 above)
Comment (peterebden) [reply src/fs/glob.go]: Ah yes, this changed a bit over time, I think earlier it did used to return two slices.

File: src/fs/glob.go [hunk 3]
```diff
@@ -174,10 +240,26 @@ func (globber *Globber) walkDir(rootPath string) (walkedDir, error) {
...
+// globPrefix extracts the fixed prefix part from a glob expression.
+// i.e. "src/fs/**/*.go" -> ["src", "fs"], ["**", "/*.go"]
+func (globber *Globber) globPrefix(expression string) []string {
+	parts := splitPath(expression)
+	for i, part := range parts {
+		if strings.ContainsAny(part, "*?[") {
+			return parts[:i]
+		}
+	}
+	// If we get here, the entire expression was fixed; leave the last item off which saves some craziness later.
```
[the remark this answers is not in the document]
Comment (peterebden) [reply src/fs/glob.go]: Ha I forget a little of the detail but will expand it a bit - I think it's just that it's easier not to handle the expression going to filepath.Glob being empty.

**Review by peterebden — COMMENTED** [review]
> This looks good! I was wondering if you got any metrics of this on the core repo?

No, but from another repo; the original message describes an order of magnitude speedup. That was a case that is fairly pathological to the current implementation; it downloads a big subrepo then does a lot of globs into it; those globs are quite "principled" but we were still taking a very long time over them (basically doing `filepath.Match` against 40k files for every `glob` call).

File: src/fs/glob.go [hunk 4]
```diff
@@ -95,6 +156,8 @@ func (globber *Globber) Glob(rootPath string, includes, excludes []string, inclu
 	var filenames []string
 	for _, include := range includes {
 		mustBeValidGlobString(include)
+		// glob(["./x"]) is equivalent to just glob(["x"])
+		include = strings.TrimPrefix(include, "./")
```
[the remark this answers is not in the document]
Comment (Tatskaari) [reply src/fs/glob.go:161]: Yeah I think that makes sense

**Comment by peterebden** [comment]
Someone is welcome to resurrect this if they think it worth it - unfortunately I don't have time to fix whatever's wrong atm.
